[2단계 - 로또 구현] 라이언(권영훈) 미션 제출합니다.

YounghoonKwon opened the pull request:
안녕하세요 ! 재연링

벌써 한주가 끝나가네요.
보람찬 한주가 되었길 바라며
2단계 미션 제출 합니다.


질문 사항

1.
수동구매 로또와 자동구매 로또를 합쳐주어야 하는 상황이었는데요,
원래의 로직은,
LottoTickets 객체 생성시, 생성자에 List<LottoTicket>을 넣어주어서
LottoTickets에서 한번에 validation 후 객체 생성 과정이었습니다.
그러나 이 과정을 하게 되면 수동구매의 자동구매 로또를 합칠 때 문제가 있었습니다.

그래서 처음 시도하였던 해결방안은, default생성자를 사용하고, 하나씩 add를 해나가자.
이거였습니다만, 이렇게 하는 경우 생성자에서 validation하는 부분을 활용하지 못하게 되는점을 감안하여,

수동구매와 자동구매의 객체를 각각 생성 후에,
그 두개의 객체를 합쳐주는 방법을 사용하여서 수정하였는데 이것이 옳은 방법인지 모르겠습니다.

2. 
LottoTicketsService의 20번째 줄에 InputView.getUserInput이 들어가게 되었는데..
최대한 컨트롤러에서 InputView를 call하게끔 로직을 짜려고 했으나
이부분은 어떻게 처리를 해야 할지 에매했네요.
서비스에 InputView를 사용하여도 괜찮은지 궁금합니다.


감사합니다! 좋은 주말 되세요!

Commit:
refactor: LottoTicketService, LottoNumber의 CACHE를 미리 섞어서 만들때 바로 받는다

Commit:
LottoWinnerTicket, LottoWinnerBonusNumber 삭제
역할의 차이가 없다. 책임이 동일하다.
변수 이름으로 충분히 구분히 가능하다.
따라서 분리할 이유가 없다

Commit:
refactor: LottoRank에 bonusMatch 필드 생성, comparator 리팩토링

Commit:
refactor: LottoRank 언더바를 이용한 가독성 향상

Commit:
refactor: LottoRank의 match함수 리팩토링

Commit:
refactor: 승자 로또 생성 재활용하기 위해 네이밍 수정

Commit:
refactor: 사용하지 않는 param 삭제

Commit:
feat: 수동 구매 로직 완료

Commit:
feat: OutputView 완성

Commit:
doc: README 업데이트

jaeyeonling commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -34,7 +36,12 @@ public int getNumber() {
     }
 
     public static List<LottoNumber> getCache() {
```
메서드명이 해당 기능의 의미를 나타내고 있을까요?
또한 랜덤한 요소가 해당 위치에 존재한다면 어떠한 문제가 있을지 고민해보아요

jaeyeonling commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -32,34 +34,47 @@ private Money initMoney() {
+        LottoTickets manualLottoTickets = initManualLottoTickets(money);
+        LottoTickets autoLottoTickets = LottoTicketsService.createAutoLottoTickets(money);
+        OutputView.printHowManyTicketsBought(manualLottoTickets.size(), autoLottoTickets.size());
+        LottoTickets mergedLottoTickets = LottoTicketsService.mergeLottoTickets(manualLottoTickets, autoLottoTickets);
```
```suggestion
        LottoTickets mergedLottoTickets = LottoTickets.merge(manualLottoTickets, autoLottoTickets);
```
```suggestion
        LottoTickets mergedLottoTickets = manualLottoTickets.merge(autoLottoTickets);
```

기존 코드와 위 두 코드는 어떠한 차이가 있을까요?

YounghoonKwon replied:
LottoTickets 에 메세지를 던지도록 구조를 바꾸어, 데이터를 가지는 객체가 일을 하도록 바뀐거 같습니다!
여기서 궁금한 점이, 1번과 같이 static 메서드를 사용할 필요가 있는 것인가? 제 생각에는 자주 쓰이는 메서드가 아니기 때문에 지양해도 될 것 같습니다. 
또 한편으로 드는 생각은, LottoTicketsService에서 새로운 객체를 만들어, merge를 하면 어떨까 하는 것입니다. 아래 두 메서드 포함 기존의 제 메서드 모두 기존 객체의 데이터가 변경되는데, 혹시 프로그램이 커짐에 따라, 두개를 통합한 객체, 수동로또 객체, 자동로또 객체, 이렇게 3개의 객체가 필요로 하는 순간이 올지도 모른다고 생각이 됩니다.
```
1. LottoTickets mergedLottoTickets = LottoTickets.merge(manualLottoTickets, autoLottoTickets);
2. LottoTickets mergedLottoTickets = manualLottoTickets.merge(autoLottoTickets);
```

jaeyeonling commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -4,13 +4,15 @@
     private static final int MAX_NUMBER_RANGE = 45;
     private static final String RANGE_ERROR_MESSAGE =
             "숫자는 + " + MIN_NUMBER_RANGE + "~" + MAX_NUMBER_RANGE + " 사이의 숫자여야한다.";
+    private static final int LOTTO_NUMBER_COUNT = LottoTicket.LOTTO_NUMBER_COUNT;
```
동일한 네이밍이라면 static import를 통해 동일한 효과를 보실 수 있습니다!

jaeyeonling commented on src/main/java/lotto/domain/LottoTicket.java:
```diff
@@ -34,7 +34,14 @@ public LottoRank getRank(LottoWinner lottoWinner) {
+        if (lottoNumbers.contains(lottoWinnerBonusNumber)) {
+            return true;
+        }
+        return false;
```
```suggestion
        return lottoNumbers.contains(lottoWinnerBonusNumber);
```

와 같이 작성이 가능하겠네요!

jaeyeonling commented on src/main/java/lotto/service/LottoTicketsService.java:
```diff
@@ -13,11 +14,27 @@ private LottoTicketsService() {
-            lottoTicketGroup.add(LottoTicketService.createLottoTicket());
+        for (int i = 0; i < manualTicketsCount; i++) {
+            lottoTicketGroup.add(LottoTicketService.createManualLottoTicket(InputView.getUserInput()));
         }
```
> LottoTicketsService의 20번째 줄에 InputView.getUserInput이 들어가게 되었는데..
최대한 컨트롤러에서 InputView를 call하게끔 로직을 짜려고 했으나
이부분은 어떻게 처리를 해야 할지 에매했네요.
서비스에 InputView를 사용하여도 괜찮은지 궁금합니다.

Service의 로직과 View의 로직은 분리해주시는 것이 맞습니다
해당 위치에서 View를 통해 받는 것이 아닌 View에서 받은 데이터를 묶어 List<String>로 넘겨주면 해결 할 수 있을 것 같은데 어떻게 생각하시나요?

YounghoonKwon replied:
맞습니다!

jaeyeonling reviewed (changes requested):
2단계 미션을 잘 구현해주셨습니다 👍 
질문의 답변과 추가적인 리뷰 남겨두었으니 확인 후 반영해주세요 :)

Commit:
refactor: getCache 메서드 네이밍 수정 -> getRandomLottoNumbers

Commit:
refactor: mergeLottoTickets 메서드 객체에 메세지를 던지도록 구조 변경, 데이터를 가지는 메서드가 일을 하도록

Commit:
refactor: 상수 동일한 네이밍 static import 를 통해 처리

Commit:
refactor: contains 메서드 if 문 삭제

YounghoonKwon commented:
문득 생각난 네이밍에 대한 궁금한점이 하나 있습니다.

예를 들어, merge메서드 네이밍도
구지 mergeLottoTickets()를 하지 않고 merge() 라고 네이밍을 지어주어도 괜찮을거 같고..

Money 클래스에서 deductMoney()라는 메서드네이밍도
deduct()라고 지어도 될거 같은데..

어떤게 정답일까요?
어짜피 객체에서 의미를 전달 하기 때문에, money.deductMoney보다는 money.deduct가, lottotickets1.mergeLottoTickets(lottotickets2) 보다는 lottotickets1.merge(lottotickets2)가 의미도 충분히 전달 하면서 중복 의미 포함을 하지 않고 네이밍도 더 깔끔해지는 느낌인데... 재연링은 어떻게 생각 하시나요?

Commit:
refactor: service 로직과 view 로직 분리

jaeyeonling commented on src/main/java/lotto/service/LottoTicketsService.java:
```diff
@@ -13,11 +13,22 @@ private LottoTicketsService() {
+        List<LottoTicket> lottoTicketGroup = new ArrayList<>();
+        int moneyLeftOver = money.lottoCount();
+        for (int i = 0; i < moneyLeftOver; i++) {
+            lottoTicketGroup.add(LottoTicketService.createAutoLottoTicket());
```
https://github.com/woowacourse/java-lotto/pull/329#discussion_r584086690
에서 말한 랜덤한 요소에 대해 생각해보면 좋을 것 같아요!
해당 위치에 존재한다면 테스트를 어떻게 할 수 있을까요?
힌트로는 자동차 미션에서 움직임에 대한 테스트를 고민해보면 좋을 것 같습니다  :)

jaeyeonling commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -32,34 +34,47 @@ private Money initMoney() {
+        LottoTickets manualLottoTickets = initManualLottoTickets(money);
+        LottoTickets autoLottoTickets = LottoTicketsService.createAutoLottoTickets(money);
+        OutputView.printHowManyTicketsBought(manualLottoTickets.size(), autoLottoTickets.size());
+        LottoTickets mergedLottoTickets = manualLottoTickets.mergeLottoTickets(autoLottoTickets);
```
> 여기서 궁금한 점이, 1번과 같이 static 메서드를 사용할 필요가 있는 것인가? 제 생각에는 자주 쓰이는 메서드가 아니기 때문에 지양해도 될 것 같습니다.

개발 스타일에 따라 선택하여 구현해주시면 될 것 같습니다!

또 한편으로 드는 생각은, LottoTicketsService에서 새로운 객체를 만들어, merge를 하면 어떨까 하는 것입니다. 아래 두 메서드 포함 기존의 제 메서드 모두 기존 객체의 데이터가 변경되는데, 혹시 프로그램이 커짐에 따라, 두개를 통합한 객체, 수동로또 객체, 자동로또 객체, 이렇게 3개의 객체가 필요로 하는 순간이 올지도 모른다고 생각이 됩니다.

실수를 방지할 수 있어 대부분의 값은 불변으로 사용하는 것을 권장합니다 
저도 merge 메서드 내부에서 기존 값을 변경하지 않고, 새로운 컬렉션을 만들어 응답하게 구현할 것 같네요!

jaeyeonling commented on src/main/java/lotto/domain/LottoTickets.java:
```diff
@@ -16,10 +16,23 @@ public LottoTickets(final List<LottoTicket> lottoTickets) {
         this.lottoTickets = lottoTickets;
     }
 
+    public LottoTickets mergeLottoTickets(LottoTickets otherLottoTickets) {
```
>문득 생각난 네이밍에 대한 궁금한점이 하나 있습니다.
>
>예를 들어, merge메서드 네이밍도
구지 mergeLottoTickets()를 하지 않고 merge() 라고 네이밍을 지어주어도 괜찮을거 같고..
>
>Money 클래스에서 deductMoney()라는 메서드네이밍도
deduct()라고 지어도 될거 같은데..
>
>어떤게 정답일까요?
어짜피 객체에서 의미를 전달 하기 때문에, money.deductMoney보다는 money.deduct가, lottotickets1.mergeLottoTickets(lottotickets2) 보다는 lottotickets1.merge(lottotickets2)가 의미도 충분히 전달 하면서 중복 의미 포함을 하지 않고 네이밍도 더 깔끔해지는 느낌인데... 재연링은 어떻게 생각 하시나요?

저도 네이밍의 중복을 제거하는 방식으로 구현하는 것을 추천합니다!
상황에 따라 명시해주는 것이  더 가독성이 높아지는 경우도 있지만 위 예시의 경우 제거를 해주시는 것이 더 가독성이 올라갈 것 같습니다 :)

jaeyeonling reviewed (changes requested):
피드백 잘 반영해주셨습니다!
질문의 답변과 피드백 남겨두었으니 확인해주세요~

Commit:
refactor: LottoTicket 자동 생성시, 테스트를 할수 있게끔 List<LottoNumber>을 받아오게끔 수정

Commit:
refactor: 전체적인 네이밍 리팩토링

Commit:
refactor: merge 메서드 새로운 컬랙션 생성 후 리턴

Commit:
refactor: LottoTicketService 네이밍 및 함수 리팩토링

Commit:
feat: LottoTicketService 테스트 추가

YounghoonKwon commented:
https://github.com/woowacourse/java-lotto/pull/329#discussion_r584261721

여기서 새로운 컬렉션을 만들어 응답하는 것도, 데이터를 가지는 객체가 일을 하게끔 하는 것이 맞는 것인가요?
도메인에서 새로운 컬렉션을 만들어 응답해본적이 없어서 그런지,
뭔가 데이터를 가지는 객체 내에서 새로운 컬렉션을 만들어 응답하는것은 서비스에서 해줘야될거 같은 느낌이 들어서요.

jaeyeonling commented on src/main/java/lotto/domain/LottoTickets.java:
```diff
@@ -12,15 +12,29 @@
+        LottoTickets mergedTickets = new LottoTickets(lottoTickets);
+        mergedTickets.addAll(otherLottoTickets);
+        return mergedTickets;
+    }
```
```suggestion
    public LottoTickets merge(LottoTickets otherLottoTickets) {
        final List<LottoTicket> mergedTickets = new ArrayList<>(this.lottoTickets.size() + otherLottoTickets.lottoTickets.size());
        mergedTickets.addAll(this.lottoTickets);
        mergedTickets.addAll(this.otherLottoTickets.lottoTickets);
        return mergedTickets;
    }
```

기존의 값을 변화시키지 않고 사용해주는 것이 좋을 것 같네요!
객체를 불변하게 유지한다고 생각하시면 좋을 것 같습니다 :)

YounghoonKwon replied:
앗 LottoTickets mergedTickets = new LottoTickets(lottoTickets);
이줄에서 불변객체를 생성하는것으로 생각하고 있었네요 ㅠ 제 실수입니다:)

new ArrayList<>(사이즈)
여기서 new ArrayList생성시에 사이즈를 설정해주는 것은 다른 의도가 있을까요?

jaeyeonling commented on src/main/java/lotto/service/LottoTicketsService.java:
```diff
@@ -13,11 +14,22 @@ private LottoTicketsService() {
+        List<LottoTicket> lottoTicketGroup = new ArrayList<>();
+        int moneyLeftOver = money.ticketCount();
+        for (int i = 0; i < moneyLeftOver; i++) {
+            lottoTicketGroup.add(LottoTicketService.createTicket(LottoNumber.getRandomNumbers()));
```
https://github.com/woowacourse/java-lotto/pull/329#discussion_r584260887 
해당 피드백에 대한 해결이 됐을까요!?
어려운 부분이 있으시다면 편하게 질문 주세요 :)

YounghoonKwon replied:
제가 이해 한 바로는 createTicket이란 메서드 내에서 랜덤번호 생성하는 것이 포함되있었는데
그것을 분리 하라는 뜻으로 이해했어요!

혹시 제가 잘못 짚은 걸까요...?

기존코드
```
        List<LottoTicket> lottoTicketGroup = new ArrayList<>();
        int moneyLeftOver = money.lottoCount();
        for (int i = 0; i < moneyLeftOver; i++) {
            lottoTicketGroup.add(LottoTicketService.createAutoLottoTicket());
```
변경 후
```
        List<LottoTicket> lottoTicketGroup = new ArrayList<>();
        int moneyLeftOver = money.ticketCount();
        for (int i = 0; i < moneyLeftOver; i++) {
            lottoTicketGroup.add(LottoTicketService.createTicket(LottoNumber.getRandomNumbers()));
```

jaeyeonling reviewed (changes requested):
피드백 잘 반영해주셨습니다!
소소한 피드백 남겨두었으니 확인 부탁드릴게요 :)

Commit:
refactor: LottoTicket들을 merge 시킬 때, 새로운 불변 객체 생성 후 반환

Commit:
refactor: 랜덤 로또 티켓 생성 인터페이스 분리

Commit:
fix: Money 관련 버그 픽스, 복사 생성자 구현
로또 티켓을 만들 때마다 현재 갖고 있는 돈에서 티켓을 만든만큼 돈을 빼왔었는데,
그때문에 결과 출력에 오류가 있었음. deepcopy를 이용해 수정(clonable 쓰는거보다 생성자 복사가 좋다고함)

Commit:
refactor: 사용 않는 import 제거

Commit:
refactor: 사용하지 않는 메서드 제거

Commit:
fix: 갖고 있는 돈만큼 수동 구매시 에러 발생 수정

jaeyeonling reviewed (approved):
추가적인 요구사항도 잘 반영해주셨습니다 👍 
수 많은 피드백 반영하시느라 고생 많으셨어요
다음 단계도 화이팅입니다 :)
